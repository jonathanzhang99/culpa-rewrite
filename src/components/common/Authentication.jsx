import PropTypes from "prop-types";
import React, { useContext, useReducer } from "react";
import { Route, Redirect } from "react-router-dom";

export const AuthContext = React.createContext(null);

function authReducer(state, action) {
  switch (action.type) {
    case "LOGIN_START":
      return {
        ...state,
        isLoading: true,
      };
    case "LOGIN_ERROR":
      return {
        ...state,
        isError: true,
        isLoading: false,
      };

    case "LOGIN_SUCCESS":
      localStorage.setItem("authToken", JSON.stringify(action.payload));
      return {
        ...state,
        isError: false,
        isLoading: false,
        isLoggedIn: true,
        authToken: action.payload,
      };

    case "LOGIN_FAILURE":
      return {
        ...state,
        isError: false,
        isLoading: false,
        isLoggedIn: false,
      };
    case "LOGOUT_SUCCESS":
      localStorage.removeItem("authToken");
      return {
        ...state,
        isError: false,
        isLoading: false,
        isLoggedIn: false,
        authToken: undefined,
      };
    default:
      throw new Error();
  }
}

const propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]).isRequired,
};

/**
 * AuthProvider createa React Context with values such that children elements
 * in the tree can access user auth token and manipulate them
 * (e.g. login, logout)
 *
 */
export function AuthProvider({ children }) {
  const existingToken = JSON.parse(localStorage.getItem("authToken"));

  const [state, dispatch] = useReducer(authReducer, {
    isError: false,
    isLoading: false,
    isLoggedIn: !!existingToken,
    authToken: existingToken,
  });

  // `login` either returns an error or null if a token was succesfully
  // retrieved. The error will always be a string message
  const login = async ({ username, password }) => {
    dispatch({ type: "LOGIN_START" });
    const response = await fetch("/api/auth/login", {
      method: "POST",
      body: JSON.stringify({
        username,
        password,
      }),
      headers: {
        "Content-Type": "application/json",
      },
    });
    try {
      const result = await response.json();

      // Require that the backend sends either a token or an error
      result.error
        ? dispatch({ type: "LOGIN_FAILURE" })
        : dispatch({ type: "LOGIN_SUCCESS", payload: result.token });
      return result;
    } catch (err) {
      dispatch({ type: "LOGIN_ERROR" });
      return err;
    }
  };

  // `logout` simply removes the token from the client. It does NOT
  // expire the token.
  const logout = async () => {
    if (state.authToken) {
      dispatch({ type: "LOGOUT_SUCCESS" });
    }
  };

  return (
    <AuthContext.Provider
      value={{
        ...state,
        login,
        logout,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
}

AuthProvider.propTypes = propTypes;

/**
 * ProtectedRoute allows access to the url only if a user is logged in.
 *
 */
export function ProtectedRoute(props) {
  const { isLoggedIn } = useContext(AuthContext);
  // eslint-disable-next-line react/jsx-props-no-spreading
  return isLoggedIn ? <Route {...props} /> : <Redirect to="/" />;
}
