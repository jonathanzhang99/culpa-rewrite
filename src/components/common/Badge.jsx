import PropTypes from "prop-types";
import React from "react";
import { Image } from "semantic-ui-react";

import bronzeNugget from "icons/bronze-nugget.svg";
import goldNugget from "icons/gold-nugget.svg";
import silverNugget from "icons/silver-nugget.svg";

const propTypesBadge = {
  badgeId: PropTypes.number.isRequired,
};

export default function Badge({ badgeId }) {
  switch (badgeId) {
    case 1: {
      return <Image avatar spaced="left" src={goldNugget} />;
    }
    case 2: {
      return <Image avatar spaced="left" src={silverNugget} />;
    }
    case 3: {
      return <Image avatar spaced="left" src={bronzeNugget} />;
    }
    default: {
      return null;
    }
  }
}

Badge.propTypes = propTypesBadge;
