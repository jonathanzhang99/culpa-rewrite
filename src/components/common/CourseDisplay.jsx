import PropTypes from "prop-types";
import React from "react";
import { Link } from "react-router-dom";
import { Header } from "semantic-ui-react";

const propTypesCourseDisplayName = {
  as: PropTypes.oneOf(["header", "text"]),
  courseCallNumber: PropTypes.string,
  courseName: PropTypes.string.isRequired,
  size: PropTypes.oneOf(["tiny", "small", "medium", "large", "huge"]),
};

const propTypesCourseDisplayLink = {
  as: PropTypes.oneOf(["header", "text"]),
  courseCallNumber: PropTypes.string,
  courseId: PropTypes.number.isRequired,
  courseName: PropTypes.string.isRequired,
  size: PropTypes.oneOf(["tiny", "small", "medium", "large", "huge"]),
};

const defaultProps = {
  courseCallNumber: "",
  as: "text",
  size: "medium",
};

export function CourseDisplayName({ as, courseCallNumber, courseName, size }) {
  const CourseName = as === "header" ? Header : "span";
  return (
    <CourseName size={size}>
      {courseCallNumber} {courseName}
    </CourseName>
  );
}

export function CourseDisplayLink({
  as,
  courseCallNumber,
  courseId,
  courseName,
  size,
}) {
  return (
    <Link to={`/course/${courseId}`}>
      <CourseDisplayName
        as={as}
        courseCallNumber={courseCallNumber}
        courseName={courseName}
        size={size}
      />
    </Link>
  );
}

CourseDisplayName.propTypes = propTypesCourseDisplayName;
CourseDisplayName.defaultProps = defaultProps;

CourseDisplayLink.propTypes = propTypesCourseDisplayLink;
CourseDisplayLink.defaultProps = defaultProps;
