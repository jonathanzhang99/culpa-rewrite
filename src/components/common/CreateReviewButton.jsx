/* eslint-disable react/jsx-props-no-spreading */
import PropTypes from "prop-types";
import React from "react";
import { Link } from "react-router-dom";
import { Button } from "semantic-ui-react";

const propTypes = {
  color: PropTypes.string,
  courseId: PropTypes.number,
  fluid: PropTypes.bool,
  subject: PropTypes.string,
  professorId: PropTypes.number,
  relaxed: PropTypes.bool,
};

const defaultProps = {
  color: "orange",
  courseId: undefined,
  fluid: false,
  subject: "",
  professorId: undefined,
  relaxed: false,
};

export default function CreateReviewButton({
  color,
  courseId,
  fluid,
  professorId,
  subject,
  relaxed,
}) {
  return (
    <Link
      to={{
        pathname: "/review/submit",
        state: { courseId, professorId },
      }}
    >
      <Button
        className={`${relaxed ? "relaxed" : ""} create-review-button`}
        color={color}
        fluid={fluid}
      >
        WRITE A REVIEW {subject ? `FOR ${subject.toUpperCase()}` : ""}
      </Button>
    </Link>
  );
}

CreateReviewButton.propTypes = propTypes;
CreateReviewButton.defaultProps = defaultProps;
