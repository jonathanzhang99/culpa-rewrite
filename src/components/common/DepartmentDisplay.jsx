import PropTypes from "prop-types";
import React from "react";
import { Link } from "react-router-dom";
import { Header } from "semantic-ui-react";

const propTypesDepartmentDisplayName = {
  as: PropTypes.oneOf(["header", "text"]),
  departmentName: PropTypes.string.isRequired,
  size: PropTypes.oneOf(["tiny", "small", "medium", "large", "huge"]),
};

const propTypesDepartmentDisplayLink = {
  as: PropTypes.oneOf(["header", "text"]),
  departmentId: PropTypes.number.isRequired,
  departmentName: PropTypes.string.isRequired,
  size: PropTypes.oneOf(["tiny", "small", "medium", "large", "huge"]),
};

const defaultProps = {
  as: "text",
  size: "medium",
};

export function DepartmentDisplayName({ as, departmentName, size }) {
  const DepartmentName = as === "header" ? Header : "span";
  return <DepartmentName size={size}> {departmentName} </DepartmentName>;
}

export function DepartmentDisplayLink({
  as,
  departmentId,
  departmentName,
  size,
}) {
  return (
    <Link to={`/department/${departmentId}`}>
      <DepartmentDisplayName
        as={as}
        departmentName={departmentName}
        size={size}
      />
    </Link>
  );
}

DepartmentDisplayName.propTypes = propTypesDepartmentDisplayName;
DepartmentDisplayName.defaultProps = defaultProps;

DepartmentDisplayLink.propTypes = propTypesDepartmentDisplayLink;
DepartmentDisplayLink.defaultProps = defaultProps;
