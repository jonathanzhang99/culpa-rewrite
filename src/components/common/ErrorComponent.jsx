import React from "react";

export default function ErrorComponent() {
  return <div>error occurred</div>;
}
