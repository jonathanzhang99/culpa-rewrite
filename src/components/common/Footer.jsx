import React from "react";
import { Link } from "react-router-dom";
import { Grid, Icon, Image } from "semantic-ui-react";

import culpaIcon from "icons/culpa.svg";

export default function Footer() {
  return (
    <Grid className="footer" columns={2} verticalAlign="middle">
      <Grid.Column textAlign="center">
        <Image centered src={culpaIcon} />
        <Icon color="blue" name="copyright outline" size="small" />
        <span style={{ color: "#82B7E8" }}>1997-2020</span>
      </Grid.Column>
      <Grid.Column textAlign="left">
        <Grid.Row>
          <Link className="footer-link" to="/">
            About
          </Link>
        </Grid.Row>
        <Grid.Row>
          <Link className="footer-link" to="/join-the-team">
            Join The Team
          </Link>
        </Grid.Row>
        <Grid.Row>
          <Icon color="blue" name="mail" size="small" />
          <span className="footer-link">admin@culpa.info</span>
        </Grid.Row>
      </Grid.Column>
    </Grid>
  );
}
