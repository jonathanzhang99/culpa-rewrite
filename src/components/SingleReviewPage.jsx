import PropTypes from "prop-types";
import React from "react";
import { useParams, Link } from "react-router-dom";
import { Container, Header, Button } from "semantic-ui-react";

import ErrorComponent from "components/common/ErrorComponent";
import LoadingComponent from "components/common/LoadingComponent";
import useDataFetch from "components/common/useDataFetch";
import ReviewCard from "components/reviews/ReviewCard";

const propTypesThankYouTextBox = {
  reviewId: PropTypes.number.isRequired,
};

export function ThankYouTextBox({ reviewId }) {
  return (
    <Container fluid style={{ margin: "20vh" }} textAlign="center">
      <Header as="h1">Thank you!</Header>
      <Header
        as="h3"
        style={{
          lineHeight: "30px",
          margin: "30px",
        }}
      >
        Please save your
        <span style={{ color: "#004E8D" }}> Review ID: {reviewId} </span>
        in case you need to refer to your submission in the future.
      </Header>
      <Link to="/review/submit">
        <Button color="orange" size="medium">
          WRITE ANOTHER REVIEW
        </Button>
      </Link>
      <Link to="/">
        <Button color="blue" size="medium">
          BACK TO HOMEPAGE
        </Button>
      </Link>
    </Container>
  );
}

ThankYouTextBox.propTypes = propTypesThankYouTextBox;

export default function SingleReviewPage() {
  const { reviewId } = useParams();
  const {
    data: { review, flag },
    isLoading,
    isError,
  } = useDataFetch(`/api/review/${reviewId}`, {
    review: {},
    flag: "",
  });

  if (isLoading || isError) {
    return isLoading ? <LoadingComponent /> : <ErrorComponent />;
  }

  return (
    <div>
      {flag === "approved" ? (
        <ReviewCard
          content={review.content}
          deprecated={review.deprecated}
          rating={review.rating}
          reviewHeader={review.reviewHeader}
          reviewId={review.reviewId}
          reviewType={review.reviewType}
          submissionDate={review.submissionDate}
          votes={review.votes}
          workload={review.workload}
        />
      ) : (
        review.reviewId && <ThankYouTextBox reviewId={review.reviewId} />
      )}
    </div>
  );
}
