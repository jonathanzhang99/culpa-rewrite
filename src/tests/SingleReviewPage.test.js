import { act, render, screen } from "@testing-library/react";
import React from "react";
import { MemoryRouter, Route } from "react-router-dom";

import { AuthProvider } from "components/common/Authentication";
import SingleReviewPage, { ThankYouTextBox } from "components/SingleReviewPage";

describe("single review page", () => {
  const cases = [
    {
      reviewId: 1,
      fetchReturn: {
        flag: "approved",
        review: {
          reviewId: 1,
          reviewType: "all",
          reviewHeader: {
            course: {
              courseId: 1,
              courseName: "test course",
              courseCode: "ABCD 1234",
            },
            professor: {
              profId: 1,
              profFirstName: "Jane",
              profLastName: "Doe",
              uni: "jd6789",
              badges: [1],
            },
          },
          votes: {
            initUpvoteCount: 1,
            initDownvoteCount: 2,
            initFunnyCount: 3,
            upvoteClicked: true,
            downvoteClicked: false,
            funnyClicked: false,
          },
          content: "demo demo",
          workload: "demo demo",
          rating: 3,
          submissionDate: "Oct 13, 2018",
          deprecated: false,
        },
      },
    },
    {
      reviewId: 2,
      fetchReturn: {
        flag: "libel",
        review: {
          reviewId: 2,
        },
      },
    },
    {
      reviewId: 3,
      fetchReturn: {
        flag: "pending",
        review: {
          reviewId: 3,
        },
      },
    },
  ];

  cases.forEach(({ reviewId, fetchReturn }) => {
    test(`${fetchReturn.flag} case logic test`, async () => {
      const mockFetch = jest.spyOn(global, "fetch");
      mockFetch.mockImplementation(() =>
        Promise.resolve({
          ok: true,
          json: () => fetchReturn,
        })
      );

      await act(async () => {
        render(
          <MemoryRouter initialEntries={[`/review/${reviewId}`]}>
            <Route path="/review/:reviewId">
              <AuthProvider>
                <SingleReviewPage />
              </AuthProvider>
            </Route>
          </MemoryRouter>
        );
      });

      expect(mockFetch).toHaveBeenCalled();

      if (fetchReturn.flag === "approved") {
        expect(screen.getByText("Workload")).toBeInTheDocument();
        expect(screen.getByText(`ID: ${reviewId}`)).toBeInTheDocument();
      } else {
        expect(screen.getByText("Thank you!")).toBeInTheDocument();
        expect(screen.getByText("WRITE ANOTHER REVIEW")).toBeInTheDocument();
        expect(screen.getByText(`Review ID: ${reviewId}`)).toBeInTheDocument();
      }
    });
  });
});

describe("Thank you text box", () => {
  test("snapshot test", () => {
    const snapshot = render(
      <MemoryRouter>
        <ThankYouTextBox reviewId={12345} />
      </MemoryRouter>
    );
    expect(snapshot).toMatchSnapshot();
  });
});
