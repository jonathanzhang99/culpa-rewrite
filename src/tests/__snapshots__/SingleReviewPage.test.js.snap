// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thank you text box snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui fluid center aligned container"
        style="margin: 20vh;"
      >
        <div>
          <div
            class="ui huge header"
          >
            Thank you!
          </div>
        </div>
        <div>
          <div
            class="ui header"
          >
            Please save your
            <span
              style="color: rgb(0, 78, 141);"
            >
               Review ID: 
              12345
               
            </span>
            in case you need to refer to your submission in the future.
          </div>
        </div>
        <div>
          <a
            href="/review/submit"
          >
            <button
              class="ui orange huge button relaxed"
            >
              WRITE ANOTHER REVIEW
            </button>
          </a>
          <a
            href="/"
          >
            <button
              class="ui blue huge button relaxed"
            >
              BACK TO HOMEPAGE
            </button>
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui fluid center aligned container"
      style="margin: 20vh;"
    >
      <div>
        <div
          class="ui huge header"
        >
          Thank you!
        </div>
      </div>
      <div>
        <div
          class="ui header"
        >
          Please save your
          <span
            style="color: rgb(0, 78, 141);"
          >
             Review ID: 
            12345
             
          </span>
          in case you need to refer to your submission in the future.
        </div>
      </div>
      <div>
        <a
          href="/review/submit"
        >
          <button
            class="ui orange huge button relaxed"
          >
            WRITE ANOTHER REVIEW
          </button>
        </a>
        <a
          href="/"
        >
          <button
            class="ui blue huge button relaxed"
          >
            BACK TO HOMEPAGE
          </button>
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
